src/systems/a320_systems/src/hydraulic.rs

```diff
@@ -1479,28 +1547,217 @@ impl SimulationElement for A320BrakingForce {
     }
 }
 
-struct Door {
-    exit_id: String,
-    position: f64,
-    previous_position: f64,
+#[derive(PartialEq, Clone, Copy)]
+pub enum DoorControlState {
+    DownLocked = 0,
+    NoControl = 1,
+    HydControl = 2,
+    UpLocked = 3,
+}
+
+struct A320DoorController {
+    requested_position_id: String,
+
+    state: DoorControlState,
+
+    position_requested: Ratio,
+
+    duration_in_no_control: Duration,
+    duration_in_hyd_control: Duration,
+
+    should_close_valves: bool,
+    control_position_request: Ratio,
+    should_unlock: bool,
 }
-impl Door {
-    fn new(id: usize) -> Self {
+impl A320DoorController {
+    // Duration which the hydraulic valves sends a open request when request is closing (this is done on real aircraft so uplock can be easily unlocked without friction)
+    const UP_CONTROL_TIME_BEFORE_DOWN_CONTROL: Duration = Duration::from_millis(200);
+
+    // Delay from the ground crew unlocking the door to the time they start requiring up movement in control panel
+    const DELAY_UNLOCK_TO_HYDRAULIC_CONTROL: Duration = Duration::from_secs(5);
+
+    fn new(id: &str) -> Self {
         Self {
-            exit_id: format!("EXIT OPEN:{}", id),
-            position: 0.,
-            previous_position: 0.,
+            requested_position_id: format!("{}_DOOR_CARGO_OPEN_REQ", id),
+            state: DoorControlState::DownLocked,
+            position_requested: Ratio::new::<ratio>(0.),
+
+            duration_in_no_control: Duration::from_secs(0),
+            duration_in_hyd_control: Duration::from_secs(0),
+
+            should_close_valves: true,
+            control_position_request: Ratio::new::<ratio>(0.),
+            should_unlock: false,
         }
     }
 
-    fn has_moved(&self) -> bool {
-        (self.position - self.previous_position).abs() > f64::EPSILON
+    fn update(&mut self, context: &UpdateContext, door: &CargoDoor, pressure_available: Pressure) {
```

Comment:
**suggestion**: `pressure_available` reads like it would be a `bool` argument, but it is not. Rename to `available_pressure` to make clear you're talking about the available amount of pressure.
**note**: I would like to note that in other situations this is named `current_pressure`. I'd pick either `available_pressure` or `current_pressure` and use that throughout the code.
```suggestion
    fn update(&mut self, context: &UpdateContext, door: &CargoDoor, available_pressure: Pressure) {
```